src/sources/soundsourceproviderregistry.cpp

```diff
@@ -1,56 +1,193 @@
+    if (!pProvider) {
+        qWarning() << "Failed to obtain SoundSource provider from plugin library"
+                << pPluginLibrary->getFilePath();
+        return; // abort registration
+    }
+    const QStringList supportedFileExtensions(
+            pProvider->getSupportedFileExtensions());
+    if (supportedFileExtensions.isEmpty()) {
+        qWarning() << "SoundSource provider"
+                << pProvider->getName()
+                << "does not support any file extensions";
+        return; // abort registration
+    }
+    for (const auto& fileExt: supportedFileExtensions) {
+        SoundSourceProviderPriority providerPriority(
+                pProvider->getPriorityHint(fileExt));
+        registerPluginProviderForFileExtension(
+                fileExt,
+                pPluginLibrary,
+                pProvider,
+                providerPriority);
+    }
+}
+
+void SoundSourceProviderRegistry::registerProviderForFileExtension(
+        const QString& fileExtension,
+        const SoundSourceProviderPointer& pProvider,
+        SoundSourceProviderPriority providerPriority) {
+    SoundSourceProviderRegistration registration(
+        SoundSourcePluginLibraryPointer(), pProvider, providerPriority);
+    addRegistrationForFileExtension(fileExtension, std::move(registration));
+}
+
+void SoundSourceProviderRegistry::registerPluginProviderForFileExtension(
+        const QString& fileExtension,
+        const SoundSourcePluginLibraryPointer& pPluginLibrary,
+        const SoundSourceProviderPointer& pProvider,
+        SoundSourceProviderPriority providerPriority) {
+    SoundSourceProviderRegistration registration(
+            pPluginLibrary, pProvider, providerPriority);
+    addRegistrationForFileExtension(fileExtension, std::move(registration));
+}
+
+void SoundSourceProviderRegistry::addRegistrationForFileExtension(
+        const QString& fileExtension,
+        SoundSourceProviderRegistration registration) {
+    DEBUG_ASSERT(registration.getProvider());
+    QList<SoundSourceProviderRegistration>& registrationsForFileExtension =
+            m_registry[fileExtension];
+    QList<SoundSourceProviderRegistration>::iterator listIter(
+            registrationsForFileExtension.begin());
+    insertRegistration(&registrationsForFileExtension, registration);
+}
+
+void SoundSourceProviderRegistry::reregisterProviderForFileExtension(
+        const QString& fileExtension,
+        const SoundSourceProviderPointer& pProvider,
+        SoundSourceProviderPriority providerPriority) {
+    QList<SoundSourceProviderRegistration>& registrationsForFileExtension =
+            m_registry[fileExtension];
+    QList<SoundSourceProviderRegistration>::iterator listIter(
+            registrationsForFileExtension.begin());
+    // Perform a linear search through the list
+    while (registrationsForFileExtension.end() != listIter) {
+        if (listIter->getProvider() == pProvider) {
```

Comment:
I will delete this function since it is unused, untested, and the desired 
functionality is not clearly specified ;)

On 11/25/2015 01:54 PM, Daniel Schürmann wrote:
>
> In src/sources/soundsourceproviderregistry.cpp 
> <https://github.com/mixxxdj/mixxx/pull/605#discussion_r45863056>:
>
> > +    QList<SoundSourceProviderRegistration>::iterator listIter(
> > +            registrationsForFileExtension.begin());
> > +    insertRegistration(&registrationsForFileExtension, registration);
> > +}
> > +
> > +void SoundSourceProviderRegistry::reregisterProviderForFileExtension(
> > +        const QString& fileExtension,
> > +        const SoundSourceProviderPointer& pProvider,
> > +        SoundSourceProviderPriority providerPriority) {
> > +    QList<SoundSourceProviderRegistration>& registrationsForFileExtension =
> > +            m_registry[fileExtension];
> > +    QList<SoundSourceProviderRegistration>::iterator listIter(
> > +            registrationsForFileExtension.begin());
> > +    // Perform a linear search through the list
> > +    while (registrationsForFileExtension.end() != listIter) {
> > +        if (listIter->getProvider() == pProvider) {
>
> I do not understand why there is no else branch and how it should work.
>
> Here the provider was already registered. If the priority has changed the 
> provider is removed and registered again.
> The priority remains the old one.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/605/files#r45863056>.
>